### PR TITLE
[codex] Reduce pipeline redundancy with focused owners

### PR DESCRIPTION
**Pull request by zfifteen**
## Summary
- extract focused owners for scene validation, metadata resolution, voice-cache readiness, and render verification
- reduce duplicated shell logic in the main pipeline while preserving the canonical entrypoint
- add deterministic regression tests for the new seams and review-discovered edge cases

## Verification
- python3 scripts/test_scene_validator.py
- python3 scripts/test_build_scenes_validator_integration.py
- python3 scripts/test_scene_repair_invocation.py
- python3 scripts/test_resolve_scene_metadata.py
- python3 scripts/test_build_scenes_metadata_resolver_integration.py
- python3 scripts/test_state_record_error_integration.py
- python3 scripts/test_phase_progress_sentinel.py
- python3 scripts/test_voice_cache_integration.py
- python3 scripts/test_final_render_verification_integration.py
- python3 scripts/test_verify_scene_video.py
- python3 scripts/test_record_scene_rendered.py
- python3 scripts/test_contracts_vs_runtime.py
- bash -n scripts/build_video.sh

Replaces closed PR #112 for a fresh review surface.

### CODE REVIEW

File: scripts/build_video.sh [hunk 1]
```diff
@@ -1064,23 +1238,22 @@ validate_scene_runtime() {
...
-  voice_output_dir="$(voice_output_dir_abs)" || return 1
-  cache_index="${voice_output_dir}/cache.json"
-  if [[ -f "$cache_index" ]]; then
+  if "$PYTHON_BIN" "${SCRIPT_DIR}/ensure_voice_cache.py" --project-dir "$PROJECT_DIR" --check >/dev/null 2>&1; then
     return 0
   fi
 
   echo "→ Voice cache index missing; generating cache before runtime validation..." | tee -a "$LOG_FILE"
-  if ! $PYTHON_BIN "${SCRIPT_DIR}/precache_voiceovers_qwen.py" "$PROJECT_DIR" \
+  if ! "$PYTHON_BIN" "${SCRIPT_DIR}/precache_voiceovers_qwen.py" "$PROJECT_DIR" \
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/build_video.sh:1254]: Fixed in 2e5225d. `ensure_qwen_cache_index()` now honors `--skip-precache`: if `ensure_voice_cache.py --check` fails and `SKIP_PRECACHE` is set, it logs the cache check output and returns nonzero before any call to `precache_voiceovers_qwen.py`. Added an integration assertion that the skip branch appears before the generation path.

File: scripts/build_video.sh [hunk 2]
```diff
@@ -1963,32 +2112,25 @@ PY
     return 0
   fi
 
-  narration_key="$(get_scene_narration_key "$scene_id")"
-  if [[ -z "$narration_key" ]]; then
-    narration_key="$scene_id"
-  fi
-
-  if [[ -z "$scene_id" || -z "$scene_file" || -z "$scene_class" ]]; then
+  if [[ -z "$scene_id" || -z "$scene_file" || -z "$narration_key" ]]; then
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/build_video.sh:2123]: Fixed in 2e5225d. `handle_build_scenes()` now reconciles the resolved `narration_key` through `get_scene_narration_key "$scene_id"` after metadata validation and before scaffold generation, so stale state keys are corrected against `narration_script.py` before being embedded in a new scene file. Added an integration assertion for that ordering.

File: scripts/ensure_voice_cache.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        has_narration_key = isinstance(narration_key, str) and bool(narration_key.strip())
+        has_text = bool(_cache_text(entry))
+        audio_file = _cache_audio_file(entry)
+
+        if not audio_file:
+            return f"cache entry {index} has no audio file"
+        if not has_narration_key and not has_text:
+            return f"cache entry {index} has neither narration_key nor text"
+        if not (cache_dir / audio_file).exists():
+            return f"cache entry {index} audio file missing: {audio_file}"
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/ensure_voice_cache.py:154]: Fixed in 9adc070. `ensure_voice_cache.py` now rejects absolute audio paths and resolves relative audio paths against the cache directory, failing if the resolved path escapes that directory before checking existence. Added tests for both absolute-path and `..` traversal cache entries.

File: scripts/scene_validator.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+    ]
+    if auto_adjust:
+        cmd.append("--auto-adjust")
+    result = subprocess.run(cmd, capture_output=True, text=True, check=False)
+    details = [line for line in (result.stdout + result.stderr).splitlines() if line.strip()]
+    if result.returncode == 0:
+        return _pass("timing_budget", "Timing budget validation passed", details)
+    if result.returncode == 2:
+        return _pass("timing_budget", "Timing budget validation indeterminate", details)
+    return _fail("timing_budget", "Scene failed deterministic timing budget validation", details)
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/scene_validator.py:385]: Fixed in 9adc070. `validate_scene_timing_budget.py` now resolves the timing cache directory through `voice_clone_config.json` using the shared `selected_output_dir()` helper instead of hardcoding `media/voiceovers/qwen/cache.json`. Added regression coverage proving the timing-budget gate passes when cache data lives in a configured non-default output dir.

File: scripts/validate_scene_timing_budget.py [hunk 5]
```diff
@@ -82,8 +84,25 @@ def _read_cache_entries(cache_data: Any) -> Iterable[dict[str, Any]]:
...
 
 
+def _load_voice_config(project_dir: Path) -> dict[str, Any]:
+    config_path = project_dir / "voice_clone_config.json"
+    if not config_path.exists():
+        return {}
+    data = json.loads(config_path.read_text(encoding="utf-8"))
+    if not isinstance(data, dict):
+        return {}
+    return data
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/validate_scene_timing_budget.py:98]: Fixed in 0f77038. `_load_voice_config()` now catches `OSError` and `json.JSONDecodeError`, emits a `[timing-budget] WARN:` message, and falls back to `{}` instead of crashing. Added a regression test with malformed `voice_clone_config.json` that confirms the timing-budget check uses the default cache path and exits cleanly.

File: scripts/scene_validator.py [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+
+
+def validate_voiceover_sync(scene_text: str) -> GateResult:
+    if 'voiceover(text="' in scene_text or "voiceover(text='" in scene_text:
+        return _fail(
+            "voiceover_sync",
+            "Scene uses hardcoded narration text instead of SCRIPT dictionary",
+        )
+    if 'voiceover(text=f"' in scene_text or "voiceover(text=f'" in scene_text:
+        return _fail(
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/scene_validator.py]: Fixed in 0f77038. `validate_voiceover_sync()` now parses the scene AST and inspects `self.voiceover(...)` calls structurally, requiring the `text` keyword to be a `SCRIPT[...]` subscript. Added focused tests for whitespace/multiline hardcoded narration rejection and spaced `SCRIPT[...]` acceptance.

File: scripts/scene_validator.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+            "template_structure",
+            "Scene missing required voiceover wrapper using SCRIPT key",
+        )
+    if voiceover_count > 1:
+        return _fail(
+            "template_structure",
+            f"Scene has nested/duplicate voiceover wrappers ({voiceover_count} found)",
+        )
+
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/scene_validator.py:103]: Fixed in 65b7747. `VOICEOVER_SCRIPT_RE` now tolerates whitespace and newlines around the scaffold `text = SCRIPT[...]` argument, aligning the template-structure gate with the AST-based voiceover-sync gate. Added a regression test for spaced/multiline `SCRIPT[...]` voiceover formatting.

File: scripts/ensure_voice_cache.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+    config_path = project_dir / "voice_clone_config.json"
+    if not config_path.exists():
+        return {}
+    try:
+        data = json.loads(config_path.read_text(encoding="utf-8"))
+    except json.JSONDecodeError as exc:
+        raise ValueError(f"voice_clone_config.json is not valid JSON: {exc}") from exc
+    if not isinstance(data, dict):
+        raise ValueError("voice_clone_config.json root must be an object")
+    return data
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/ensure_voice_cache.py:45]: Fixed in c8789b6. `load_voice_config()` now catches `OSError` separately and returns a structured `ValueError` through `check_voice_cache()` instead of allowing a traceback. I also added regression coverage for `voice_clone_config.json` read failures and cache-index read failures so the owner keeps stable JSON failure reporting.

File: scripts/scene_validator.py [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+    if not body_lines:
+        return _fail("scene_body_contract", "scene_body must contain a non-comment statement")
+
+    try:
+        for token in tokenize.generate_tokens(io.StringIO(candidate).readline):
+            if token.type == tokenize.COMMENT:
+                return _fail(
+                    "scene_body_contract",
+                    f"scene_body line {token.start[0]} must not contain comments",
+                )
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/scene_validator.py:238]: No code change here. The comment rejection is intentional in the current first-pass scene contract: `harness_responses/prompts/build_scenes/user.md` explicitly forbids comments, `harness_responses/parser.py` rejects `tokenize.COMMENT`, and `scripts/test_contracts_vs_runtime.py` pins that runtime contract. The earlier non-comment body check is only there to distinguish empty/comment-only bodies before the stricter no-comments gate runs.

File: scripts/verify_scene_video.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        return trimmed
+    return trimmed[: limit - 3] + "..."
+
+
+def verify_scene_video(project_dir: Path, scene_id: str, class_name: str) -> dict[str, Any]:
+    video_path = scene_video_path(project_dir, scene_id, class_name)
+    if not video_path.exists():
+        return _result(False, video_path, "render output missing", False, None)
+    if video_path.stat().st_size <= 0:
+        return _result(False, video_path, "render output empty", False, None)
```
[the remark this answers is not in the document]
Comment (zfifteen) [reply scripts/verify_scene_video.py]: Fixed in c8789b6. `verify_scene_video()` now wraps the render-output existence/stat checks and `ffprobe` execution in deterministic error handling, returning `_result(ok=False, ...)` for `OSError` instead of raising. Added focused tests for stat failures and ffprobe execution failures.
